DOCSP-12158: Fundamentals Documents (#39)
mongodb/docs-java: source/fundamentals/data-formats/documents.txt

Overview
--------

A MongoDB **document** is a data structure that contains key/value pairs in

> binary JSON (BSON) format and is the base unit in the MongoDB data hierarchy.

Author: I'll replace this with something similar.

Each **document** belongs to a single **collection** and each **collection**

> belongs to a single **database**. For more information, see our MongoDB manual

Author: Thanks for bringing this up.

1) I think I could replace this with `:term:` (see Glossary terms in https://docs.mongodb.com/meta/style-guide/style/text-formatting/).
Italics could be used for introducing new terms (see "New terms" in https://docs.mongodb.com/meta/style-guide/style/text-formatting/ ) explained immediately afterwards

2) I did think about creating a graphic and searched through server/manual documentation and wasn't able to find anything suitable to link to or re-use. Maybe I can't find the right page which probably means our users can't find it either. Instead of attempting to explain it from scratch, I just wanted to introduce repetition of terms and what they are as a learning aid.

That said, if this was a language-specific concept, I would probably devote a section for it.

> belongs to a single **database**. For more information, see our MongoDB manual

Reviewer: I think this is a perfect opportunity for creating a glossary, so definitely think about that! Just check to make sure it's supported properly by nextgen first :)

I hoped that the linked server page would include such a graphic, too, but sadly it doesn't. Might be a good idea to ask in the #docs chat to see if anybody knows of such a graphic (or would be willing to create one for you, if they have a LucidChart license and some degree of artistic skills)

> belongs to a single **database**. For more information, see our MongoDB manual

Reviewer: +1 to repetition as a learning aid, I think that works quite well here.

> belongs to a single **database**. For more information, see our MongoDB manual

Author: Didn't realize you need a glossary to use `:term:`. I removed formatting for the time being. I'm a little confused by the inconsistent formatting here, so probably something to address sometime later: https://github.com/mongodb/docs-java/blob/master/source/fundamentals/databases-collections.txt

entry on :manual:`Documents </core/document>`.

The Java driver and BSON library include the following document classes that
help you access and manipulate the BSON data in documents:

.. list-table::
   :header-rows: 1
   :stub-columns: 1
   :widths: 10 10 20 20

   * - Name
     - Package
     - Implements Map
     - Recommended Usage

   * - ``Document``
     - ``org.bson``
     - Yes, implements ``Map<String, Object>``
     - When you want a flexible and concise data representation.

   * - ``BsonDocument``
     - ``org.bson``
     - Yes, implements ``Map<String, BsonValue>``
     - When you need a type-safe API.

   * - ``BasicDBObject``
     - ``com.mongodb``
     - No

>      - When you are migrating your application from a legacy driver version.

Author: Thanks, I wonder if this should be a pattern we reuse for pages that present multiple similar topics.

While you can use any of the document classes in your application, we
recommend that you use the ``Document`` class since it can concisely
represent dynamically structured documents of any complexity. It
implements the ``Map<String, Object>`` interface which enables it to use
loosely-typed values.

.. warning:: Duplicate Field Names

   Although a document is valid if it contains two duplicate keys, the driver

>    only sets the value of the key to the value in the last instance.

Author: I don't have any qualms with adding this info, but "field" is included in the linked doc: https://docs.mongodb.com/manual/core/document/#field-names

I'll make sure I explain that the link has important info about field-related terminology and limitations, and remove this warning. I think the reason it is called out in the current version of the driver docs is because they don't link to the manual entry: https://mongodb.github.io/mongo-java-driver/4.1/bson/documents/

Document
--------

The ``Document`` class offers a flexible representation of a BSON document.

> You can access and manipulate document data using Java types with this

Reviewer: ```suggestion
You can access and manipulate fields using Java types with this
```
Maybe this is a good opportunity to use the "field" term to familiarize readers? (see previous comment)

Author: Appending the fix to another commit.

> class. See the following table for mappings between frequently-used BSON

Reviewer: nit: should "frequently-used" be "frequently used"? I tried to look it up in the meta style guide but I couldn't find it there.

Author: I think usually you use hyphen for adverb + past participle verb, but maybe open for debate.
There's some discussion here:
https://english.stackexchange.com/questions/225433/should-i-used-a-hyphen-for-often-used-and-well-documented

I'm going to leave it _as-is_ for now, unless there's a pressing need to change it.

Reviewer: Sounds good to me!

and Java types:

.. list-table::
   :header-rows: 1
   :stub-columns: 1

   * - BSON type
     - Java type
   * - Array
     - ``java.util.List``
   * - Binary
     - ``org.bson.types.Binary``
   * - Boolean
     - ``java.lang.Boolean``
   * - Date
     - ``java.util.Date``
   * - Document
     - ``org.bson.Document``
   * - Double
     - ``java.lang.Double``
   * - Int32
     - ``java.lang.Integer``
   * - Int64
     - ``java.lang.Long``
   * - Null
     - ``null``

>    * - ObjectId``

Reviewer: ```suggestion
   * - ObjectId
```
I think you accidentally a monospace!

Author: Whoops, good catch. Even the autobuilder did not warn me about that.

     - ``org.bson.types.ObjectId``
   * - String
     - ``java.lang.String``

> You can also customize the type mapping by specifying a custom codec.

Reviewer: Nit: It seems like "also" is unnecessary here. What do you think?
```suggestion
You can customize the type mapping by specifying a custom codec.
```

Author: I was also waffling with the wording of this sentence. I think I might need to provide more detail as I was hoping I could keep it concise and that the reader could infer that the table showed default mappings and that they could override them by specifying or implementing a custom codec.

For more information on customizing mapped types, see our guide on using
:doc:`Codecs </fundamentals/data-formats/codecs>`.

In the following code snippet, we show how to instantiate and build a sample

> ``Document`` instance that contains a variety of data types:

Reviewer: Thought: Is "contains a variety of data types" the best descriptor we can offer here? I wonder if this glosses over the fact that a document actually contains a collection of string keys that map to values with a variety of data types. Perhaps we can add some more detail about that fact here (and, my favorite bone to pick at this point, add the term "field" here)?

Author: I think you might have been expecting a description of a Document class. I was actually attempting to communicate that the example intentionally uses field values of different types to demonstrate how to set them. I'll clarify that here. 

Regarding data structures contained within the Document class: I want to limit the scope of what is discussed in reference documentation since the reader can refer to API documentation for those details.

.. code-block:: java
   :copyable: true

   Document author = new Document("_id", new ObjectId())
       .append("name", "Gabriel García Márquez")
       .append("lastUpdated", new Date(120, 2, 26))
       .append("novels", Arrays.asList(
           new Document("title", "One Hundred Years of Solitude").append("yearPublished", 1967),
           new Document("title", "Chronicle of a Death Foretold").append("yearPublished", 1981),
           new Document("title", "Love in the Time of Cholera").append("yearPublished", 1985)));

You can insert this document into a collection using the following code:

> .. code-block:: java

Reviewer: For some reason, syntax highlighting doesn't seem to be working for this example.

   :copyable: true

>   // MongoClient mongoClient = ...

Author: It was designed to shows the type with the instantiation as the variable is used later on. I can certainly add more explanation.

>   // MongoClient mongoClient = ...

Reviewer: I think that your solution is solid here -- it gives just enough of a hint that a reader can google for a solution (or search on the docs, or navigate within the docs) if they don't already know how to instantiate a mongoClient, mongoCollection, etc.

>   MongoDatabase database = mongoClient.getDatabase("test_data");

Reviewer: Because Document is the default representation for database documents, it seems like you're just relying on the defaults here. Should we call this out explicitly above?

Reviewer: The equivalent code snippets for other document representations use the "test_fundamentals" database instead of "test_data". Should we standardize the database used?

Author: > Because Document is the default representation for database documents, it seems like you're just relying on the defaults here. Should we call this out explicitly above?

I'm guessing you meant this on the next line for `getCollection`. I'll add a mention and link to the API docs.

  MongoCollection<Document> collection = database.getCollection("authors");

  InsertOneResult result = collection.insertOne(author);

Once you perform a successful insert, you can retrieve the sample document
data from the collection using the following code:

.. code-block:: java
   :copyable: true

   import com.mongodb.client.model.Filters;

>    ...

Author: From [this commen](https://github.com/mongodb/docs-java/pull/39#discussion_r527092414)t, I need to add some explanation of the dots.

   Document doc = collection.find(Filters.eq("novels.yearPublished", 1967)).first();
   if (doc != null) {
       System.out.println("_id: " + doc.getObjectId("_id")
           + ", name: " + doc.getString("name")
           + ", lastUpdated: " + doc.getDate("lastUpdated"));

       doc.getList("novels", Document.class).forEach((novel) -> {
           System.out.println("title: " + novel.getString("title")
               + ", yearPublished: " + novel.getInteger("yearPublished"));
           });

>    }

Author: >  include a sentence or two clarifying how you access field values from an Instance of Document

Could you clarify what you mean by this?  I included methods called on the Document instance to access the fields such as getObjectId, getString, and getDate. I also showed how to access other types within an embedded Document.

For the second part:
>what happens if you use the wrong type of get (for instance, getString on an ObjectId field)

I'll think about it how I might incorporate that info.

>    }

Reviewer: I very much like the note that you added after these examples -- that's precisely what I was suggesting above. Resolved!

The output should resemble the following:

.. code-block:: none
   :copyable: false

   _id: 5fb5fad05f734e3794741a35, name: Gabriel García Márquez, lastUpdated: Thu Mar 26 00:00:00 EDT 2020
   title: One Hundred Years of Solitude, yearPublished: 1967
   title: Chronicle of a Death Foretold, yearPublished: 1981
   title: Love in the Time of Cholera, yearPublished: 1985

For more information on retrieving and manipulating MongoDB data, see our
:doc:`CRUD guide </fundamentals/crud>`.

For more information on this class, see the

> :java-docs:`Document API documentation </apidocs/bson/org/bson/Document.html>`.

Author: Thanks for the suggestion. I think you mean `seealso`. I've been avoiding using it, but I think it doesn't really change the experience either way. I'll keep it in mind though since the change in formatting it produces could be a good signal to readers if used consistently.

BsonDocument
------------

The ``BsonDocument`` class provides a type-safe API to access and manipulate
a BSON document. You need to specify the BSON type from the Java BSON
library for each field. See the following table for mappings between

> frequently-used BSON and Java BSON library types:

Reviewer: duplicate of an earlier nit: should "frequently-used" be "frequently used"? I tried to look it up in the meta style guide but I couldn't find it there.

.. list-table::
   :header-rows: 1
   :stub-columns: 1

   * - BSON type
     - Java BSON library type
   * - Array
     - ``org.bson.BsonArray``
   * - Binary
     - ``org.bson.BsonBinary``
   * - Boolean
     - ``org.bson.Boolean``
   * - Date (long value)
     - ``org.bson.BsonDateTime``
   * - Document
     - ``org.bson.BsonDocument``
   * - Double
     - ``org.bson.BsonDouble``
   * - Int32
     - ``org.bson.BsonInt32``
   * - Int64
     - ``org.bson.BsonInt64``
   * - Null
     - ``null``
   * - ObjectId
     - ``org.bson.BsonObjectId``
   * - String
     - ``org.bson.BsonString``

> You can also customize the type mapping by specifying a custom codec.

Reviewer: ```suggestion
You can customize the type mapping by specifying a custom codec.
```
Nit: It seems to me like "also" is unnecessary here. What do you think?

For more information on customizing mapped types, see our guide on using
:doc:`Codecs </fundamentals/data-formats/codecs>`.

In the following code snippet, we show how to instantiate and build a sample
``BsonDocument`` instance that contains a variety of data types:

.. code-block:: java
   :copyable: true

   BsonDocument author = new BsonDocument()
       .append("_id", new BsonObjectId())
       .append("name", new BsonString("Gabriel García Márquez"))
       .append("lastUpdated", new BsonDateTime(new Date(120, 2, 26).getTime()))
       .append("novels", new BsonArray(Arrays.asList(
           new BsonDocument().append("title", new BsonString("One Hundred Years of Solitude")).append("yearPublished", new BsonInt32(1967)),
           new BsonDocument().append("title", new BsonString("Chronicle of a Death Foretold")).append("yearPublished", new BsonInt32(1981)),
           new BsonDocument().append("title", new BsonString("Love in the Time of Cholera")).append("yearPublished", new BsonInt32(1985))
      )));

You can insert this document into a collection using the following code:

.. code-block:: java

>    :copyable: true

Reviewer: While you have specified that these blocks should be copyable, I don't see a copy button in your staged version. Nextgen bug?

Author: Will investigate. When I last checked, the first code blocks I created and previewed were working fine, but maybe I introduced some syntax issue along the way

   // MongoClient mongoClient = ...

   MongoCollection<BsonDocument> collection = database.getCollection("authors", BsonDocument.class);

>    MongoDatabase database = mongoClient.getDatabase("fundamentals_data");

Reviewer: I might be going crazy but I *think* these lines are in the wrong order!

Author: Yeah, good copypasta catch.

   InsertOneResult result = collection.insertOne(author);

Once you perform a successful insert, you can retrieve the sample document
data from the collection using the following code:

.. code-block:: java
   :copyable: true

   import com.mongodb.client.model.Filters;
   ...

>    BsonDocument doc = collection.find(Filters.eq("novels.yearPublished", 1967)).first();

Reviewer: I wonder if this query could confuse users who aren't hugely familiar with MongoDB, since we're using some fairly advanced ideas here (querying an embedded document with dot syntax, matching a document based off a match to one field in a single document in a collection). On one hand, this does demonstrate some interesting MongoDB edge cases, and a lot of readers might be curious about dot syntax. On the other hand... when I started out with MongoDB, I might have had some trouble following this. Perhaps it would be better to use a simpler filter here and instead cover these more advanced concepts in the CRUD section?

   if (doc != null) {
       System.out.println("_id: " + doc.getObjectId("_id").getValue()
               + ", name: " + doc.getString("name").getValue()
               + ", lastUpdated: " + new Date(doc.getDateTime("lastUpdated").getValue()));

       doc.getArray("novels").forEach((novel) -> {
           System.out.println("title: " + novel.asDocument().getString("title").getValue()
                   + ", yearPublished: " + novel.asDocument().getInt32("yearPublished").getValue());
       });
   }

The output should resemble the following:

.. code-block:: none
   :copyable: false

   _id: 5fb5fad05f734e3794741a35, name: Gabriel García Márquez, lastUpdated: Thu Mar 26 00:00:00 EDT 2020
   title: One Hundred Years of Solitude, yearPublished: 1967
   title: Chronicle of a Death Foretold, yearPublished: 1981
   title: Love in the Time of Cholera, yearPublished: 1985

For more information on retrieving and manipulating MongoDB data, see our
:doc:`CRUD guide </fundamentals/crud>`.

> For more information on this class, see the

Reviewer: Duplicate: perhaps a good use for a see also directive?

:java-docs:`BsonDocument API documentation </apidocs/bson/org/bson/BsonDocument.html>`.

BasicDBObject
-------------

The ``BasicDBObject`` class allows you to access and manipulate document data
using Java types. We recommend that you avoid using this class unless you
are migrating an application from that uses an older driver version. In

> addition to deprecation, this class has the following limitations:

Reviewer: Instead of the mildly awkward "in addition to deprecation" phrasing, I wonder if we could:
1. Add (Deprecated) after the section header, akin to the (BETA) label for Realm Sync
2. Mention "deprecated" in the first sentence ("The deprecated ``BasicDBObject`` class") to make the deprecation even more obvious.

We could then just change this sentence to "This class has the following limitations" and, if we really want to hammer in the deprecation idea, add "This class is deprecated" as a new first bullet point in the limitations.

Author: Thought it was deprecated, but I guess not.

- This class implements the ``DBObject`` interface, so you cannot extend its

>   API without breaking binary compatibility.

Reviewer: Perhaps I'm just not familiar with "binary compatibility" or API extension constraints, but it's unclear to me how the first clause relates to the second clause (I'm assuming a causal relationship). It could also be helpful to define "binary compatibility" here (are we talking about binaries as in programs? Or binary representation of objects in process memory?

Author: I copied this from the existing docs, but I'm assuming it is referring to https://wiki.eclipse.org/Evolving_Java-based_APIs_2#Evolving_API_packages -- see example 4. (A more recent doc: https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html )

As I understand it, "extend its API" means add a method to a class that implements it such as BasicDBObject. Suppose you had a compiled app that uses driver v4.1, but v4.2 came out with a new method, BasicDBObject#coolNewMethod. Normally, you could just upgrade the jar in the classpath and restart the app (or hot-reload) and it would work fine.  However, since you broke binary compatibility with v4.2 by adding a method, you need to recompile your app for it to work with v4.2. However, if there was some non-breaking change such as a private method change, your app would continue to work without recompilation because it's not visible to your binary.

Admittedly, it is a little unclear which object the pronoun "its" refers to so I can reword. It's also probably less frequently encountered than the next item, so I'll re-order it in the list.

.. code-block:: java
   :copyable: true

   // MongoClient mongoClient = ...

   MongoDatabase database = mongoClient.getDatabase("fundamentals_data");
   MongoCollection<BasicDBObject> collection = database.getCollection("authors", BasicDBObject.class);

   InsertOneResult result = collection.insertOne(author);

Once you perform a successful insert, you can retrieve the sample document
data from the collection using the following code:

.. code-block:: java
   :copyable: true

   import com.mongodb.client.model.Filters;
   ...

   BasicDBObject doc = collection.find(Filters.eq("novels.yearPublished", 1967)).first();
   if (doc != null) {
       System.out.println("_id: " + doc.getObjectId("_id")
           + ", name: " + doc.getString("name")
           + ", lastUpdated: " + doc.getDate("lastUpdated"));

       BasicDBList novels = (BasicDBList) doc.get("novels");
       if (novels != null) {
           BasicDBObject[] novelArr = novels.toArray(new BasicDBObject[0]);
           for (BasicDBObject novel : novelArr) {
               System.out.println("title: " + novel.getString("title")
               + ", yearPublished: " + novel.getInt("yearPublished"));
           }
       }

>    }

Reviewer: Similar thoughts as I had for the Document class -- should we explain the usage and constraints of the BasicDBObject type getters?

Author: Addressed in a note directive

The output should resemble the following:

.. code-block:: none
   :copyable: false

   _id: 5fb5fad05f734e3794741a35, name: Gabriel García Márquez, lastUpdated: Thu Mar 26 00:00:00 EDT 2020
   title: One Hundred Years of Solitude, yearPublished: 1967
   title: Chronicle of a Death Foretold, yearPublished: 1981
   title: Love in the Time of Cholera, yearPublished: 1985

For more information on retrieving and manipulating MongoDB data, see our
:doc:`CRUD guide </fundamentals/crud>`.

For more information on this class, see the
:java-docs:`BasicDBObject API documentation </apidocs/mongodb-driver-core/com/mongodb/BasicDBObject.html>`.

Summary
-------

In this guide, we covered the following topics on Java document classes:

- Described three Java classes you can use to work with MongoDB documents and
  why you might prefer one over the other.
- Provided usage examples for each class on building documents containing
  multiple types, inserting them into a collection, and

>   retrieving/accessing their typed fields.

Reviewer: Nice summary, I really like this.

Author: Thanks!

Overview
--------

A MongoDB **document** is a data structure that contains key/value pairs in
binary JSON (BSON) format. The document is the base unit of the MongoDB

> data hierarchy: each document belongs to a single collection

Reviewer 2: > The document is the base unit of the MongoDB	data hierarchy

I feel like you're inventing new terminology here: I've never heard a document described this way in our documentation.  And it's not accurate.  The base unit is a BSON _value_ (e.g. int64, int32, boolean), not a BSON document.

And also, documents are used for other purposed besides storage in collections: e.g. commands, queries, sort criteria, etc

Author: Good point, I'll remove the phrase that mentions it's the base unit.

read our page on :manual:`Documents </core/document>` in the MongoDB manual.

The Java driver and BSON library include the following document classes that
help you access and manipulate the BSON data in documents:

.. list-table::
   :header-rows: 1
   :stub-columns: 1
   :widths: 10 10 20 20

   * - Name
     - Package
     - Implements Map
     - Recommended Usage

   * - ``Document``
     - ``org.bson``
     - Yes, implements ``Map<String, Object>``
     - When you want a flexible and concise data representation.

   * - ``BsonDocument``
     - ``org.bson``
     - Yes, implements ``Map<String, BsonValue>``
     - When you need a type-safe API.

   * - ``BasicDBObject``
     - ``com.mongodb``
     - No
     - When you are migrating your application from a legacy driver version.

While you can use any of the document classes in your application, we
recommend that you use the ``Document`` class since it can concisely
represent dynamically structured documents of any complexity. It
implements the ``Map<String, Object>`` interface which enables it to use
loosely-typed values.

Document
--------

The ``Document`` class offers a flexible representation of a BSON document.

> You can access and manipulate fields using Java types with this class. See

Reviewer 2: "using Java types from the standard library"  would be more accurate.  BsonDocument also uses Java types, but ones that are defined in our own bson library

Author: Thanks, will update.

the following table for mappings between frequently-used BSON and Java types:

.. list-table::
   :header-rows: 1
   :stub-columns: 1

   * - BSON type
     - Java type
   * - Array
     - ``java.util.List``
   * - Binary
     - ``org.bson.types.Binary``
   * - Boolean
     - ``java.lang.Boolean``
   * - Date
     - ``java.util.Date``
   * - Document
     - ``org.bson.Document``
   * - Double
     - ``java.lang.Double``
   * - Int32
     - ``java.lang.Integer``
   * - Int64
     - ``java.lang.Long``
   * - Null
     - ``null``
   * - ObjectId
     - ``org.bson.types.ObjectId``
   * - String
     - ``java.lang.String``

The mapping table above shows the default mapping when working with the
``Document`` class. You can customize the type mapping by specifying a
custom codec. For more information on customizing mapped types, see our guide
on using :doc:`Codecs </fundamentals/data-formats/codecs>`.

In the following code snippet, we show how to instantiate and build a sample
``Document`` instance representing a document containing several
different field types:

.. code-block:: java

   Document author = new Document("_id", new ObjectId())

>        .append("name", "Gabriel García Márquez")

Reviewer 3: Nice author reference!

>        .append("lastUpdated", new Date(120, 2, 26))

Reviewer 2: That's a strange date... The year 120?  

Also, that constructor is deprecated.  Horribly, the way to do this in Java is:

    Date.from(LocalDate.of(2000, 1, 1).atStartOfDay(ZoneId.systemDefault()).toInstant())

But since you're just modeling a lastUpdated field, why not just `new Date())`?

Author: Thanks, haven't kept up with Java 8+ dates. I'll change the field name and purpose since I want to show an example of setting a specific date.

  InsertOneResult result = collection.insertOne(author);

Once you perform a successful insert, you can retrieve the sample document
data from the collection using the following code:

.. code-block:: java

   import com.mongodb.client.model.Filters;
   // <MongoCollection setup code here>

   Document doc = collection.find(Filters.eq("name", "Gabriel García Márquez")).first();
   if (doc != null) {
       System.out.println("_id: " + doc.getObjectId("_id")
           + ", name: " + doc.getString("name")
           + ", lastUpdated: " + doc.getDate("lastUpdated"));

       doc.getList("novels", Document.class).forEach((novel) -> {
           System.out.println("title: " + novel.getString("title")
               + ", yearPublished: " + novel.getInteger("yearPublished"));
           });
   }

.. note::

   The code sample above uses helper methods that check the returned type
   and throw an exception if it is unable to cast the field value.
   You can call the :java-docs:`get() </apidocs/bson/org/bson/Document.html#get(java.lang.Object)>`
   method specified by the ``Map`` interface to retrieve field values as
   type ``Object`` and to skip type checking.

> The output should resemble the following:

Reviewer 3: "will" instead of "should"?

   title: One Hundred Years of Solitude, yearPublished: 1967
   title: Chronicle of a Death Foretold, yearPublished: 1981
   title: Love in the Time of Cholera, yearPublished: 1985

For more information on retrieving and manipulating MongoDB data, see our
:doc:`CRUD guide </fundamentals/crud>`.

For more information on this class, see the
:java-docs:`Document API documentation </apidocs/bson/org/bson/Document.html>`.

BsonDocument
------------

The ``BsonDocument`` class provides a type-safe API to access and manipulate
a BSON document. You need to specify the BSON type from the Java BSON
library for each field. See the following table for mappings between
frequently-used BSON and Java BSON library types:

.. list-table::
   :header-rows: 1
   :stub-columns: 1

   * - BSON type
     - Java BSON library type
   * - Array
     - ``org.bson.BsonArray``
   * - Binary
     - ``org.bson.BsonBinary``
   * - Boolean
     - ``org.bson.Boolean``
   * - Date (long value)
     - ``org.bson.BsonDateTime``
   * - Document
     - ``org.bson.BsonDocument``
   * - Double
     - ``org.bson.BsonDouble``
   * - Int32
     - ``org.bson.BsonInt32``
   * - Int64
     - ``org.bson.BsonInt64``
   * - Null

>      - ``null``

Reviewer 2: `null` is actually not allowed.  Use the `org.bson.BsonNull` type

Author: Thanks, didn't see that type. Will update.

   * - ObjectId
     - ``org.bson.BsonObjectId``
   * - String
     - ``org.bson.BsonString``

> The mapping table above shows the default mapping when working with the

Reviewer 2: You should not really customize the mapping for BsonDocument.  Let's remove this paragraph.

Author: Will remove.

``BsonDocument`` class. You can customize the type mapping by specifying a
custom codec. For more information on customizing mapped types, see our guide
on using :doc:`Codecs </fundamentals/data-formats/codecs>`.

In the following code snippet, we show how to instantiate and build a sample
``BsonDocument`` instance representing a document containing several
different field types:

.. code-block:: java

   BsonDocument author = new BsonDocument()
       .append("_id", new BsonObjectId())
       .append("name", new BsonString("Gabriel García Márquez"))

>        .append("lastUpdated", new BsonDateTime(new Date(120, 2, 26).getTime()))

Reviewer 2: As above: just use `new Date()`

Author: Will update for new field.

   // <MongoCollection setup code here>

   BsonDocument doc = collection.find(Filters.eq("name", "Gabriel García Márquez")).first();
   if (doc != null) {
       System.out.println("_id: " + doc.getObjectId("_id").getValue()
               + ", name: " + doc.getString("name").getValue()
               + ", lastUpdated: " + new Date(doc.getDateTime("lastUpdated").getValue()));

       doc.getArray("novels").forEach((novel) -> {
           System.out.println("title: " + novel.asDocument().getString("title").getValue()
                   + ", yearPublished: " + novel.asDocument().getInt32("yearPublished").getValue());
       });
   }

.. note::

   The code sample above uses helper methods that check the returned type
   and throw a :java-docs:`BsonInvalidOperationException </apidocs/bson/org/bson/BsonInvalidOperationException.html>`
   if it is unable to cast the field value. You can call the
   :java-docs:`get() </apidocs/bson/org/bson/BsonDocument.html#get(java.lang.Object)>`
   method specified by the ``Map`` interface to retrieve field values as
   type :java-docs:`BsonValue </apidocs/bson/org/bson/BsonValue.html>` and to
   skip type checking.

> The output should resemble the following:

Reviewer 3: "will" instead of "should"?

.. code-block:: none
   :copyable: false

>    _id: 5fb5fad05f734e3794741a35, name: Gabriel García Márquez, lastUpdated: Thu Mar 26 00:00:00 EDT 2020

Reviewer 2: I'm surprised you got 2020 given the date you used above

Author: Will update for new field.

BasicDBObject
-------------

The ``BasicDBObject`` class allows you to access and manipulate document data
using Java types. We recommend that you avoid using this class unless you
are migrating an application from that uses an older driver version
because of the following limitations:

- ``BasicDBObject`` does not implement ``Map<K, V>`` and therefore
  lacks the data access and manipulation convenience methods of a ``Map``.
- It implements the ``DBObject`` interface rather than a class, so you
  cannot extend the API without breaking `binary compatibility <https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html#jls-13.2>`__.
  This means that if the interface was ever changed in a way that broke
  the binary compatibility, all applications and classes that use the
  interface need to be recompiled to run the new version without errors.

In the following code snippet, we show how to instantiate and build a sample
``BasicDBObject`` instance representing a document containing several
different field types:

.. code-block:: java

   BasicDBObject author = new BasicDBObject("_id", new ObjectId())
       .append("name", "Gabriel García Márquez")

>        .append("lastUpdated", new Date(120, 2, 26))

Reviewer 2: As above

Author: Will update for new field.

data from the collection using the following code:

.. code-block:: java

   import com.mongodb.client.model.Filters;
   // <MongoCollection setup code here>

   BasicDBObject doc = collection.find(Filters.eq("name", "Gabriel García Márquez")).first();
   if (doc != null) {
       System.out.println("_id: " + doc.getObjectId("_id")
           + ", name: " + doc.getString("name")
           + ", lastUpdated: " + doc.getDate("lastUpdated"));

       BasicDBList novels = (BasicDBList) doc.get("novels");
       if (novels != null) {
           BasicDBObject[] novelArr = novels.toArray(new BasicDBObject[0]);
           for (BasicDBObject novel : novelArr) {
               System.out.println("title: " + novel.getString("title")
               + ", yearPublished: " + novel.getInt("yearPublished"));
           }
       }
   }

.. note::

   The code sample above uses helper methods that check the returned type
   and throw an exception if it is unable to cast the field value.
   You can call the :java-docs:`get() </apidocs/bson/org/bson/BasicBSONObject.html?is-external=true#get(java.lang.String)>`
   method to retrieve values as type type ``Object`` and to skip type checking.

> The output should resemble the following:

Reviewer 3: "will" instead of "should"?

Author: I avoid using "will" across docs because it often leads to passive voice and I think we prefer active when possible based on some feedback I received last year. Do you think that I should change it anyway?

.. code-block:: none
   :copyable: false

   _id: 5fb5fad05f734e3794741a35, name: Gabriel García Márquez, lastUpdated: Thu Mar 26 00:00:00 EDT 2020
   title: One Hundred Years of Solitude, yearPublished: 1967
   title: Chronicle of a Death Foretold, yearPublished: 1981
   title: Love in the Time of Cholera, yearPublished: 1985

For more information on retrieving and manipulating MongoDB data, see our
:doc:`CRUD guide </fundamentals/crud>`.

For more information on this class, see the
:java-docs:`BasicDBObject API documentation </apidocs/mongodb-driver-core/com/mongodb/BasicDBObject.html>`.

Summary
-------

> In this guide, we covered the following topics on Java document classes:

Reviewer 2: `Java document classes` is kind of awkward.  It's Java driver classes, not Java classes.  And they're not document classes, they're classes used to model/represent BSON documents.

Can you rephrase?

Author: Sure, found a couple other similar mentions in this PR, will update.

--------

A MongoDB **document** is a data structure that contains key/value pairs in

> binary JSON (BSON) format. Each document belongs to a single collection

Reviewer 2: I still think this is misleading since we use documents for queries, commands, etc.  Those don't belong to a collection.

Author: Ok thanks, I'll update the sentence. My intent is to build a little more context beyond the first sentence without having the user follow the link. If you think that's extraneous or if there's some other essential info you'd prefer to include instead, let me know.
